### ISSUE TITLE
Difficult to diagnose failure in credential handling from AWS Metadata Endpoint

### ISSUE THREAD

**Issue by Zordrak** [hashicorp/terraform-provider-aws#19716]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform 0.14.7
AWS Provider 3.30.0

### Background / Expected Behaviour / Actual Behaviour

We are running an extremely high volume terraform run in AWS that uses hundreds of AWS provider instantiations across multiple regions and account simultaneously. As regions are in-code optional, and providers cannot be conditionally instantiated, we must instantiated a provider for every possible provider that might be needed to operate across all accounts and regions.

We usually run these terraform executions from AWS CodeBuild.

Frequently we have come to have errors in many of our runs that we cannot reproduce on-demand, and are having great difficulty pinning down. The short of it is this error message:

Error: error configuring Terraform AWS Provider: no valid credential sources for Terraform AWS Provider found.
--
807 | &nbsp;
808 | Please see https://registry.terraform.io/providers/hashicorp/aws
809 | for more information about providing credentials.
810 | &nbsp;
811 | Error: NoCredentialProviders: no valid providers in chain. Deprecated.
812 | For verbose messaging see aws.Config.CredentialsChainVerboseErrors
813 | &nbsp;
814 | &nbsp;
815 | on provider_aws_audit.tf line 3, in provider "aws":
816 | 3: provider "aws" {

Effectively something is just exceeding the limits of what can be handled at one time and some of the provider instantiations fail complaining they couldn't find any credentials.

I have a theory that even our AWS Concierge team are struggling to help prove, which is that we are being rate limited at the AWS Metadata Endpoint where the instance profile IAM credentials are sourced for the provider.

I am struggling to even get a debug log output capture generated for this but I will add it to this ticket if I am able.

Is there any way to tune the settings at all in terms of the provider access to the endpoint? Does it back off on failure at all, or does it error on the first failure encountered? I wonder if in fact the logic is set up so that if a failure is encountered trying to get the credentials from the endpoint, it just treats it as if there are no creds to be found there and moves on to the next provider in the chain.

I appreciate this may even be an issue with the AWS Go SDK, but I'm struggling for pointers.

What can I do to help further diagnose this issue, or can anything be done with the provider to help be more tolerant of what I think is likely causing the error here?

**Comment by jack-parsons-nihp** [comment]
[anonymised_tflog.gz](https://github.com/hashicorp/terraform-provider-aws/files/6734947/anonymised_tflog.gz)
Please find attached a gzipped trace level log file where this error occurred. As above we suspect that it is being throttled at the metadata endpoint. We cannot see these calls in the trace log. Is there a way to output these or increase the SDK debug output?

**Comment by jack-parsons-nihp** [comment]
https://github.com/hashicorp/aws-sdk-go-base/pull/58
Once this PR is merged we will be able to provide better debug information.

**Comment by gdavison** [comment]
The AWS provider has better credential caching behaviour as of v4.0. If you're still seeing this problem, please open a new issue
